### PR TITLE
cpuid crate refactoring - part 1

### PR DESCRIPTION
**Pull request by serban300**
Issue #, if available: #851 

Description of changes:

cpuid crate refactoring - part 1: use helper methods instead of bitwise operations when editing the cpuid

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: fc_util/src/bit_helper.rs [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+
+    fn write_bits(&mut self, bit_range: &BitRange, val: u32) -> &mut Self {
+        *self &= !get_mask(&bit_range);
+        *self |= val << bit_range.right;
+        self
+    }
+}
+
+#[cfg(test)]
+mod bit_helper_tests {
```
Comment (andreeaflorescu) [inline fc_util/src/bit_helper.rs]: Test modules have to be named "tests" so we can ignore them when computing the coverage.

File: fc_util/src/bit_helper.rs [hunk 2]
```diff
@@ -0,0 +1,186 @@
+/// Structure representing a range of bits in a number
```
Comment (andreeaflorescu) [inline fc_util/src/bit_helper.rs]: If we are not going to use this in other crates as well, we should have this as a module in cpuid. The plan for cpuid is to enhance it with other features that are more generic and can be used by other projects as well. This is why it's probably best not to have dependencies on Firecracker-specific crates in cpuid.

File: cpuid/src/bit_helper.rs [hunk 3]
```diff
@@ -0,0 +1,188 @@
+/// Structure representing a range of bits in a number
+///
+pub struct BitRange {
+    /// most significant bit
+    pub left: u32,
```
Comment (acatangiu) [inline cpuid/src/bit_helper.rs]: I'm missing what `left` and `right` are used for? How is 'a range of bits' represented here exactly?

**Comment by acatangiu** [comment]
@serban300 please add clarifying code comments and a more detailed description as the commit message.

I can read the original code, but I am ashamed to admit that I really can't make sense of the bit-magic this PR is proposing by just looking at the new code.

File: cpuid/src/bit_helper.rs [hunk 4]
```diff
@@ -0,0 +1,188 @@
+/// Structure representing a range of bits in a number
+///
+pub struct BitRange {
```
Comment (acatangiu) [inline cpuid/src/bit_helper.rs]: Add a better description here as well as example(s) on how to use the struct.

File: cpuid/src/bit_helper.rs [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+impl BitHelper for u32 {
+    fn set_bit(&mut self, pos: u32, val: bool) -> &mut Self {
+        *self &= !(1 << pos);
+        *self |= (val as u32) << pos;
+        self
+    }
+
+    fn read_bits(&self, bit_range: &BitRange) -> Self {
```
Comment (acatangiu) [inline cpuid/src/bit_helper.rs]: rename to `read_bits_in_range` and clearly document that it returns a **number** of type `Self` created from the exact bits in the range of the original number but off-setted to index 0.

File: cpuid/src/bit_helper.rs [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+        (self & get_mask(&bit_range)) >> bit_range.right
+    }
+
+    fn write_bits(&mut self, bit_range: &BitRange, val: Self) -> &mut Self {
```
Comment (acatangiu) [inline cpuid/src/bit_helper.rs]: rename to `write_bits_in_range`

clearly document that it shifts `val` to overlap with the bitrange and then writes.

File: cpuid/src/bit_helper.rs [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+        *self &= !get_mask(&bit_range);
+        *self |= val << bit_range.right;
```
Comment (acatangiu) [inline cpuid/src/bit_helper.rs]: You need to do some bounds checks. If `val` doesn't fit in the bit-range this function will happily write outside of it which is unexpected.

Also check if the bitrange falls within the `Self` type.

File: cpuid/src/bit_helper.rs [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+
+impl BitHelper for u32 {
+    fn set_bit(&mut self, pos: u32, val: bool) -> &mut Self {
+        *self &= !(1 << pos);
+        *self |= (val as u32) << pos;
+        self
+    }
+
+    fn read_bits(&self, bit_range: &BitRange) -> Self {
+        (self & get_mask(&bit_range)) >> bit_range.right
```
Comment (acatangiu) [inline cpuid/src/bit_helper.rs]: Would be nice to not allow bitranges that don't fit the `Self` type size.

File: cpuid/src/cpu_leaf.rs [hunk 9]
```diff
@@ -43,18 +52,18 @@ pub mod leaf_0x1 {
         pub const PDCM_SHIFT: u32 = 15;
         // 18 = DCA Direct Cache Access (prefetch data from a memory mapped device)
         pub const MOVBE_SHIFT: u32 = 22;
-        pub const TSC_DEADLINE_TIMER_SHIFT: u32 = 24;
+        pub const TSC_DEADLINE_TIMER: u32 = 24;
         pub const OSXSAVE_SHIFT: u32 = 27;
         // Cpu is running on a hypervisor.
-        pub const HYPERVISOR_SHIFT: u32 = 31;
+        pub const HYPERVISOR: u32 = 31;
```
Comment (acatangiu) [inline cpuid/src/cpu_leaf.rs]: Also, it becomes confusing to know which are just indexes and which are ranges. `HYPERVISOR` here is an index, but `APICID` for example is a range.

What do you think about adding a name suffix for easier identification?

File: cpuid/src/lib.rs [hunk 10]
```diff
@@ -154,65 +147,62 @@ pub fn filter_cpuid(
...
-                            if ht_enabled {
-                                // When HT is enabled, there are 2 logical cores at this level
-                                // To get the next level APIC ID, shift right with 1 because we have
-                                // maximum 2 hyperthreads per core that can be represented by 1 bit.
-                                entry.eax = 1;
-                                entry.ebx = 2;
+                        // To get the next level APIC ID, shift right with at most 1 because we have
+                        // maximum 2 hyperthreads per core that can be represented by 1 bit.
+                        entry
+                            .eax
```
Comment (acatangiu) [inline cpuid/src/lib.rs]: Previous version of the code had `eax = (cpu_count > 1 && ht_enabled) as u32` which means basically `1` or `0` whereas new version has `eax.write_bits(&eax::APICID, (cpu_count > 1 && ht_enabled) as u32)` which writes `1` or `0` but at the `APICID` offset.

Why is this change necessary?

File: cpuid/src/lib.rs [hunk 11]
```diff
@@ -154,65 +147,62 @@ pub fn filter_cpuid(
...
+                            .write_bits(&eax::APICID, (cpu_count > 1 && ht_enabled) as u32);
+                        // When cpu_count ==1 or HT is disabled, there is 1 logical core at this level
+                        // Otherwise there are 2
+                        entry.ebx.write_bits(
+                            &ebx::NUM_LOGICAL_PROCESSORS,
```
Comment (acatangiu) [inline cpuid/src/lib.rs]: Same here, values `1` or `2` are written in `ebx` at `ebx::NUM_LOGICAL_PROCESSORS` offset whereas before they were at offset `0`.

File: cpuid/src/lib.rs [hunk 12]
```diff
@@ -154,65 +147,62 @@ pub fn filter_cpuid(
...
-                        }
+                        });
                     }
+                    // Core Level Processor Topology; index = 1
                     1 => {
-                        // Core Level Processor Topology; index = 1
-                        entry.eax = LEAFBH_INDEX1_APICID_SHIFT;
+                        entry
+                            .eax
+                            .write_bits(&eax::APICID, LEAFBH_INDEX1_APICID_SHIFT);
```
Comment (acatangiu) [inline cpuid/src/lib.rs]: nit: `LEAFBH_INDEX1_APICID_SHIFT` is not a `_SHIFT` here

File: cpuid/src/lib.rs [hunk 13]
```diff
@@ -154,65 +147,62 @@ pub fn filter_cpuid(
...
-                        // and it is added here as a "re-enforcement" in case we run on
-                        // different hardware
-                        entry.eax = 0;
-                        entry.ebx = 0;
                         entry.ecx = level;
                     }
                 }
-                // EDX bits 31..0 contain x2APIC ID of current logical processor
-                // x2APIC increases the size of the APIC ID from 8 bits to 32 bits
-                entry.edx = cpu_id as u32;
```
Comment (acatangiu) [inline cpuid/src/lib.rs]: `EDX` no longer populated?

File: cpuid/src/bit_helper.rs [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+    /// Reads the value stored within the specified range of bits
+    ///
+    fn read_bits(&self, bit_range: &BitRange) -> Self;
+
+    /// Stores a value within the specified range of bits
+    ///
+    fn write_bits(&mut self, bit_range: &BitRange, val: Self) -> &mut Self;
+}
+
+fn get_mask(bit_range: &BitRange) -> u32 {
```
[the remark this answers is not in the document]
Comment (serban300) [reply cpuid/src/bit_helper.rs]: Done

File: cpuid/src/bit_helper.rs (hunk 3 above)
Comment (serban300) [reply cpuid/src/bit_helper.rs]: Done

File: cpuid/src/bit_helper.rs (hunk 4 above)
Comment (serban300) [reply cpuid/src/bit_helper.rs]: Done

File: cpuid/src/bit_helper.rs (hunk 5 above)
Comment (serban300) [reply cpuid/src/bit_helper.rs]: Done

File: cpuid/src/bit_helper.rs (hunk 7 above)
Comment (serban300) [reply cpuid/src/bit_helper.rs]: Done

File: cpuid/src/bit_helper.rs (hunk 8 above)
Comment (serban300) [reply cpuid/src/bit_helper.rs]: Done

File: cpuid/src/cpu_leaf.rs [hunk 15]
```diff
@@ -12,12 +13,20 @@ pub mod leaf_0x1 {
     }
 
     pub mod ebx {
+        use bit_helper::BitRange;
+
         // The (fixed) default APIC ID.
-        pub const APICID_SHIFT: u32 = 24;
+        pub const APICID: BitRange = BitRange {
```
[the remark this answers is not in the document]
Comment (serban300) [reply cpuid/src/cpu_leaf.rs]: Done

File: cpuid/src/cpu_leaf.rs [hunk 16]
```diff
@@ -43,18 +55,18 @@ pub mod leaf_0x1 {
         pub const PDCM_SHIFT: u32 = 15;
         // 18 = DCA Direct Cache Access (prefetch data from a memory mapped device)
         pub const MOVBE_SHIFT: u32 = 22;
-        pub const TSC_DEADLINE_TIMER_SHIFT: u32 = 24;
+        pub const TSC_DEADLINE_TIMER_BITINDEX: u32 = 24;
```
Comment (dhrgit) [inline cpuid/src/cpu_leaf.rs]: Looks like now we have two different suffices for the same thing (i.e. `BITINDEX` and `SHIFT`). Looks a bit confusing.

File: cpuid/src/bit_helper.rs [hunk 17]
```diff
@@ -0,0 +1,617 @@
...
+///     <td bgcolor="green">7</td>
+///     <td bgcolor="green">6</td>
+///     <td bgcolor="green">5</td>
+///     <td bgcolor="green">4</td>
+///     <td bgcolor="green">3</td>
+///     <td>2</td>
+///     <td>1</td>
+///     <td>0</td>
+///   </tr>
+/// </table>
```
[the remark this answers is not in the document]
Comment (acatangiu) [reply cpuid/src/bit_helper.rs]: I agree it's hard to read this in the source file.

[the remark this answers is not in the document]
Comment (serban300) [reply cpuid/src/bit_helper.rs]: I use an ASCII table as discussed offline
